lib/index.js

```diff
@@ -1,44 +1,124 @@
-'use strict';
+'use strict'
 
-var Confit = require('confit');
-var CronJob = require('cron').CronJob;
-var Express = require('express');
-var Middleware = require('./middleware');
-var Path = require('path');
-var Routes = require('./routes');
-var URL = require('url');
-var Utils = require('./utils');
-var Winston = require('winston');
+const Confit = require('confit')
+const express = require('express')
+const Middleware = require('./middleware')
+const Path = require('path')
+const Routes = require('./routes')
+const URL = require('url')
+const Utils = require('./utils')
+const Winston = require('winston')
+const medUtils = require('openhim-mediator-utils')
+const mediatorConfig = require('../config/mediator')
+const Cache = require('./cache')
 
-var app = Express();
+var server
 
-// Load the config
-Confit(Path.join(__dirname, '..', 'config')).create(function(err, config) {
-  if (err) {
-    throw err;
-  }
+function setupApp (config, appConfig) {
+  const app = express()
 
   // Make the config accessible to all routes
-  Object.defineProperty(app.locals, 'config', {value: config});
-
-  Winston.clear();
-  Winston.add(Winston.transports.Console, {timestamp: true, level: config.get('logging:level')});
+  Object.defineProperty(app.locals, 'config', { value: config })
 
-  var url = URL.resolve(config.get('dhis:url'), config.get('dhis:path'));
-  // Populate the cache on startup
-  Utils.populateCache(url);
-  // Repopulate the cache based on the cron pattern
-  new CronJob(config.get('cron:pattern'), Utils.populateCache.bind(null, url), true, config.get('cron:timezone'));
+  Winston.clear()
+  Winston.add(Winston.transports.Console, { timestamp: true, level: appConfig.get('logging:level') })
 
   // Set up the routes
-  app.get('/heartbeat', Routes.uptime);
-  app.use(Middleware.stats(config.get('statsd')));
-  app.get(config.get('dhis:path'), Routes.facilityCodeLookup);
+  app.get('/heartbeat', Routes.uptime)
+  app.use(Middleware.stats(appConfig.get('statsd')))
+
+  config.routes.forEach((route) => {
+    var url = URL.resolve(route.dhisUrl, route.dhisPath)
+
+    // Populate cache on startup
+    Utils.populateCache(url)
+    Utils.addCronJob(route.cronPattern, route.cronTimezone, url)
+    Utils.addRoute(app, route.dhisPath)
+  })
+  return app
+}
+exports.setupApp = setupApp
+
+// Load the config
+function start (callback) {
+  setupConfig((err, appConfig) => {
+    if (err) {
+      callback(err)
+    }
+
+    let apiConf = appConfig.get('openhim')
+    let port = appConfig.get('server:port')
+
+    if (apiConf.api.trustSelfSigned) {
+      process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
```

Comment:
This is bad! It disables certificate verification for all requests that are made, not just the ones to the OpenHIM. This means that connections to DHIS2 could be MITM'd. We should only be applying this config to the connections to the OpenHIM by using the `rejectUnauthorized` option on the requests.